bittide-instances/src/Bittide/Instances/Hitl/HwCcTopologies.hs

```diff
@@ -488,54 +632,141 @@ hwCcTopologyTest refClkDiff sysClkDiff syncIn rxns rxps miso =
 
 tests :: HitlTestsWithPostProcData TestConfig SimConf
 tests = Map.fromList
-  [ testTopology   diamond
-  , testTopology $ complete d3
-  , testTopology $ cyclic d5
-  , testTopology $ torus2d d2 d3
-  , testTopology $ star d7
-  , testTopology $ line d4
-  , testTopology $ hourglass d3
+  [ -- CALIBRATION --
+    -----------------
+
+    -- detect the natual clock offsets to be elided from the later tests
+    calibrateClockOffsets
```

Comment:
What I don't like about the idea is that you will always need clock control to execute the initial clock synchronization, which is also what we like to test afterwards. Thus, if you wanna modify clock control at some point in the future and suddenly nothing works any more it is hard to tell whether it already failed during the (hidden) initial synchronization or during the actual test.

With the current approach the calibration is transparent, as it is also plotted as the first entry in the report, which makes it much easier to confirm that everything worked out as expected. The only real downside with the current approach is that you need to re-calibrate if you change step size, which just means you need to organize your tests accordingly and I don't see how this will lead to major problems in the future yet. Also calibrating only once safes time, which gets relevant if you plan to test 100 or 1000 parameter sets at some point.

Reply:
> The only real downside with the current approach is that you need to re-calibrate if you change step size

By just a quick scan of the code, you're not rerunning calibration right? You're just converting units?

On:

> It assumes clock offsets are static over all the runs. While we've observed something like that, we never really tested it seriously

Could we run a clean calibration at the very end and test whether the values we obtain match the ones we obtain at the very start? If we can pull that off consistently I'd feel much more confident about relying the assumption.

Reply:
I tested that assumption explicitly once [here](https://github.com/bittide/bittide-hardware/actions/runs/8771913814), which produced nice art and made me confident enough in the first place.

Other than that, we already executed tons of HITLT tests where the initial offsets had been super stable even over month. Isn't that enough reason to put reasonable trust into that assumption. I mean, just from a logical point of view: what kind of environmental influence should be able to suddenly break that assumption without us noticing it?

Reply:
It is not a problem of implementation effort. I am more concerned about

> What I don't like about the idea is that you will always need clock control to execute the initial clock synchronization, which is also what we like to test afterwards. Thus, if you wanna modify clock control at some point in the future and suddenly nothing works any more it is hard to tell whether it already failed during the (hidden) initial synchronization or during the actual test.

As I consider this issue much more likely to produce misleading data or debugging effort on the long term.

Reply:
We really need nested threads on GitHub....

I'm fine with keeping the dependent tests after your argument. So nothing has to change in the current implementation. With:

> Is **the implementation** really that [..]

I was referring to:

> Could we run a clean calibration at the very end and test whether the values we obtain match the ones we obtain at the very start? If we can pull that off consistently I'd feel much more confident about relying on the assumption.

I.e., purely an additional test. So now it looks like:

```
measure offsets

test 0
test 1
..
test n
```

My suggestion is to solidify the assumption that the offsets stay (reasonably) static over the course of running `test{0,1,..,n}` by making it look like:


```
measure offsets

test 0
test 1
..
test n

measure offsets & compare to measured offsets in first step
```

I don't know whether we'll ever trigger a failure on that last comparison, it seems unlikely to me. But I'm also not confident enough to say that opening doors / airconditioning kicking in / hardware setup changes during long tests (or big `n`s) will **not** result in failure. Get what I'm saying?

(I'm using `test` here, but this goes for `experiment`s as well.)

Reply:
Alright, then I must have missed that point before. Yeah, this is definitely possible. It only opens the question how much deviation we allow between the initial and the last measure, as we still need to incorporate natural random noise for that.

Reply:
Yeah I'd say pick something that makes it pass 10 runs or so. We can always loosen if it trips up too often.